sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs

```diff
@@ -0,0 +1,188 @@
+        private string PartitionId { get; }
+
+        /// <summary>
+        ///   An instance of a class that implements the <see cref="IPartitionProcessor" /> interface.
+        ///   It's provided by the constructor caller and it's used to process events and errors.
+        /// </summary>
+        ///
+        private IPartitionProcessor PartitionProcessor { get; }
+
+        /// <summary>
+        ///   The set of options to use for this partition pump.
+        /// </summary>
+        ///
+        private EventProcessorOptions Options { get; }
+
+        /// <summary>
+        ///   A <see cref="CancellationTokenSource"/> instance to signal the request to cancel the current running task.
+        /// </summary>
+        ///
+        private CancellationTokenSource RunningTaskTokenSource { get; set; }
+
+        /// <summary>
+        ///   The consumer used to receive events from the Azure Event Hubs service.
+        /// </summary>
+        ///
+        private EventHubConsumer InnerConsumer { get; set; }
+
+        /// <summary>
+        ///   The running task responsible for receiving events from the Azure Event Hubs service.
+        /// </summary>
+        ///
+        private Task RunningTask { get; set; }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PartitionPump"/> class.
+        /// </summary>
+        ///
+        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
+        /// <param name="consumerGroup">The name of the consumer group this partition pump is associated with.  Events are read in the context of this group.</param>
+        /// <param name="partitionId">The identifier of the Event Hub partition this partition pump is associated with.  Events will be read only from this partition.</param>
+        /// <param name="partitionProcessor">A partition processor used to process events and errors.  Its implementation must be provided by the caller.</param>
+        /// <param name="options">The set of options to use for this partition pump.</param>
+        ///
+        internal PartitionPump(EventHubClient eventHubClient,
+                               string consumerGroup,
+                               string partitionId,
+                               IPartitionProcessor partitionProcessor,
+                               EventProcessorOptions options)
+        {
+            InnerClient = eventHubClient;
+            ConsumerGroup = consumerGroup;
+            PartitionId = partitionId;
+            PartitionProcessor = partitionProcessor;
+            Options = options;
+        }
+
+        /// <summary>
+        ///   Starts the partition pump.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Start()
+        {
+            if (RunningTask == null)
```

Comment:
Most of my thoughts from `EventProcessor` with respect to threading and synchronization issues appear to apply here as well.